### PR TITLE
Move configuration to new YGConfig and pass them down to CalculateLayout

### PR DESCRIPTION
**Pull request by woehrl01**
Move configuration to new ```YGConfig``` and pass them down to CalculateLayout. See #418 .

Adds ```YGConfigNew()``` + ```YGConfigFree```, and changed ```YGSetExperimentalFeatureEnabled``` to use the config.

New function for calculation is ```YGNodeCalculateLayoutWithConfig```.

### CODE REVIEW

File: yoga/Yoga.h [hunk 1]
```diff
@@ -219,8 +226,15 @@ YG_NODE_LAYOUT_EDGE_PROPERTY(float, Padding);
 WIN_EXPORT void YGSetLogger(YGLogger logger);
 WIN_EXPORT void YGLog(YGLogLevel level, const char *message, ...);
 
-WIN_EXPORT void YGSetExperimentalFeatureEnabled(YGExperimentalFeature feature, bool enabled);
-WIN_EXPORT bool YGIsExperimentalFeatureEnabled(YGExperimentalFeature feature);
+// YGConfig
+WIN_EXPORT YGConfigRef YGConfigNew(void);
+WIN_EXPORT void YGConfigFree(const YGConfigRef config);
+
+WIN_EXPORT void YGSetExperimentalFeatureEnabled(YGConfigRef config,
```
Comment (emilsjolander) [inline yoga/Yoga.h]: const for params please

File: yoga/Yoga.c [hunk 2]
```diff
@@ -335,6 +345,19 @@ int32_t YGNodeGetInstanceCount(void) {
   return gNodeInstanceCount;
 }
 
+YGConfigRef YGConfigNew(void) {
+  const YGConfigRef config = gYGMalloc(sizeof(YGConfig));
+  YG_ASSERT(config, "Could not allocate memory for config");
+  gNodeInstanceCount++;
```
Comment (emilsjolander) [inline yoga/Yoga.c]: remove gNodeInstanceCount++ as this is not allocating a node. I don't think we care about tracking these allocations right now.

File: yoga/Yoga.c [hunk 3]
```diff
@@ -188,6 +190,14 @@ static YGNode gYGNodeDefaults = {
         },
 };
 
+static YGConfig gYGConfigDefaults = {
+    .experimentalFeatures =
+        {
+                [YGExperimentalFeatureRounding] = false,
```
Comment (emilsjolander) [inline yoga/Yoga.c]: please uses spaces for indentation

File: yoga/Yoga.c [hunk 4]
```diff
@@ -188,6 +190,14 @@ static YGNode gYGNodeDefaults = {
         },
 };
 
+static YGConfig gYGConfigDefaults = {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: const

File: yoga/Yoga.c [hunk 5]
```diff
@@ -94,6 +94,8 @@ typedef struct YGStyle {
   float aspectRatio;
 } YGStyle;
 
+typedef struct YGConfig { bool experimentalFeatures[YGExperimentalFeatureCount + 1]; } YGConfig;
```
Comment (emilsjolander) [inline yoga/Yoga.c]: can you put this on multiple lines / run formatter?

File: yoga/Yoga.c (hunk 3 above)
Comment (woehrl01) [inline yoga/Yoga.c]: those are spaces, the formater is layouting it like that.

File: yoga/Yoga.c (hunk 5 above)
Comment (woehrl01) [inline yoga/Yoga.c]: sorry, but the layouter is formating it like that, I think we need to add a second value to layout it differently. As soon as you add ```float pointScale``` this should be layouted as desired.

Comment (emilsjolander) [inline yoga/Yoga.c]: Hehe ok. The format config still needs some tweaking i guess. No problem

File: yoga/Yoga.h [hunk 6]
```diff
@@ -219,8 +226,15 @@ YG_NODE_LAYOUT_EDGE_PROPERTY(float, Padding);
...
 WIN_EXPORT void YGLog(YGLogLevel level, const char *message, ...);
 
-WIN_EXPORT void YGSetExperimentalFeatureEnabled(YGExperimentalFeature feature, bool enabled);
-WIN_EXPORT bool YGIsExperimentalFeatureEnabled(YGExperimentalFeature feature);
+// YGConfig
+WIN_EXPORT YGConfigRef YGConfigNew(void);
+WIN_EXPORT void YGConfigFree(const YGConfigRef config);
+
+WIN_EXPORT void YGSetExperimentalFeatureEnabled(const YGConfigRef config,
+                                                YGExperimentalFeature feature,
```
Comment (emilsjolander) [inline yoga/Yoga.h]: feature and enabled const please :)

File: yoga/Yoga.c [hunk 7]
```diff
@@ -3311,14 +3356,15 @@ void YGLog(YGLogLevel level, const char *format, ...) {
...
+void YGSetExperimentalFeatureEnabled(const YGConfigRef config,
+                                     YGExperimentalFeature feature,
+                                     bool enabled) {
+  config->experimentalFeatures[feature] = enabled;
 }
 
-inline bool YGIsExperimentalFeatureEnabled(YGExperimentalFeature feature) {
-  return experimentalFeatures[feature];
+inline bool YGIsExperimentalFeatureEnabled(const YGConfigRef config,
+                                           YGExperimentalFeature feature) {
```
Comment (emilsjolander) [inline yoga/Yoga.c]: const

File: yoga/Yoga.c [hunk 8]
```diff
@@ -3311,14 +3356,15 @@ void YGLog(YGLogLevel level, const char *format, ...) {
   va_end(args);
 }
 
-static bool experimentalFeatures[YGExperimentalFeatureCount + 1];
-
-void YGSetExperimentalFeatureEnabled(YGExperimentalFeature feature, bool enabled) {
-  experimentalFeatures[feature] = enabled;
+void YGSetExperimentalFeatureEnabled(const YGConfigRef config,
+                                     YGExperimentalFeature feature,
```
Comment (emilsjolander) [inline yoga/Yoga.c]: const

**Comment by emilsjolander** [comment]
I'll most likely land this next week. It takes a while to integrate bigger changes like this into internal usages 👍  Thanks for the hard work!

**Comment by woehrl01** [comment]
@emilsjolander fully reasonable! Do you also add the other missing languages?

**Comment by emilsjolander** [comment]
Yup!

**Comment by woehrl01** [comment]
Awesome!

**Pull request closed without merging**